{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating input file for download and processing script\n",
    "\n",
    "Downloaded from [**here**](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-12916?query=E-MTAB-12916)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../../raw_data/'\n",
    "meta_data_file = 'E-MTAB-12916.sdrf.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[age]</th>\n",
       "      <th>Characteristics[developmental stage]</th>\n",
       "      <th>Characteristics[sex]</th>\n",
       "      <th>Characteristics[individual]</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Characteristics[disease]</th>\n",
       "      <th>...</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Technology Type</th>\n",
       "      <th>Comment[ENA_EXPERIMENT]</th>\n",
       "      <th>Scan Name</th>\n",
       "      <th>Comment[SUBMITTED_FILE_NAME]</th>\n",
       "      <th>Comment[ENA_RUN]</th>\n",
       "      <th>Comment[FASTQ_URI]</th>\n",
       "      <th>Comment[read_index]</th>\n",
       "      <th>Comment[read_type]</th>\n",
       "      <th>Factor Value[organism part]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ERS15408104</td>\n",
       "      <td>SAMEA113412973</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811380</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>ERR11403589</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index1</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ERS15408104</td>\n",
       "      <td>SAMEA113412973</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811380</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>ERR11403589</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index2</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ERS15408104</td>\n",
       "      <td>SAMEA113412973</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811380</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>ERR11403589</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read1</td>\n",
       "      <td>cell_barcode,umi_barcode</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ERS15408104</td>\n",
       "      <td>SAMEA113412973</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811380</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>ERR11403589</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read2</td>\n",
       "      <td>single</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>ERS15408105</td>\n",
       "      <td>SAMEA113412974</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>60 to 65</td>\n",
       "      <td>adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>D7</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811381</td>\n",
       "      <td>HCAHeart9508628_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508628_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>ERR11403590</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index1</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ERS15408149</td>\n",
       "      <td>SAMEA113413018</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>70 to 75</td>\n",
       "      <td>adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>AV13</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811425</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>ERR11403634</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read2</td>\n",
       "      <td>single</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ERS15408150</td>\n",
       "      <td>SAMEA113413019</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>70 to 75</td>\n",
       "      <td>adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>AV13</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811426</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>ERR11403635</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index1</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ERS15408150</td>\n",
       "      <td>SAMEA113413019</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>70 to 75</td>\n",
       "      <td>adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>AV13</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811426</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>ERR11403635</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index2</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ERS15408150</td>\n",
       "      <td>SAMEA113413019</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>70 to 75</td>\n",
       "      <td>adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>AV13</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811426</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>ERR11403635</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read1</td>\n",
       "      <td>cell_barcode,umi_barcode</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ERS15408150</td>\n",
       "      <td>SAMEA113413019</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>70 to 75</td>\n",
       "      <td>adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>AV13</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811426</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>ERR11403635</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read2</td>\n",
       "      <td>single</td>\n",
       "      <td>atrioventriculeft cardiac atriumr node</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>188 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0       HCAHeart9508627         ERS15408104        SAMEA113412973   \n",
       "1       HCAHeart9508627         ERS15408104        SAMEA113412973   \n",
       "2       HCAHeart9508627         ERS15408104        SAMEA113412973   \n",
       "3       HCAHeart9508627         ERS15408104        SAMEA113412973   \n",
       "4       HCAHeart9508628         ERS15408105        SAMEA113412974   \n",
       "..                  ...                 ...                   ...   \n",
       "183  HCAHeartST13189996         ERS15408149        SAMEA113413018   \n",
       "184  HCAHeartST13189997         ERS15408150        SAMEA113413019   \n",
       "185  HCAHeartST13189997         ERS15408150        SAMEA113413019   \n",
       "186  HCAHeartST13189997         ERS15408150        SAMEA113413019   \n",
       "187  HCAHeartST13189997         ERS15408150        SAMEA113413019   \n",
       "\n",
       "    Characteristics[organism] Characteristics[age]  \\\n",
       "0                Homo sapiens             55 to 60   \n",
       "1                Homo sapiens             55 to 60   \n",
       "2                Homo sapiens             55 to 60   \n",
       "3                Homo sapiens             55 to 60   \n",
       "4                Homo sapiens             60 to 65   \n",
       "..                        ...                  ...   \n",
       "183              Homo sapiens             70 to 75   \n",
       "184              Homo sapiens             70 to 75   \n",
       "185              Homo sapiens             70 to 75   \n",
       "186              Homo sapiens             70 to 75   \n",
       "187              Homo sapiens             70 to 75   \n",
       "\n",
       "    Characteristics[developmental stage] Characteristics[sex]  \\\n",
       "0                                  adult                 Male   \n",
       "1                                  adult                 Male   \n",
       "2                                  adult                 Male   \n",
       "3                                  adult                 Male   \n",
       "4                                  adult                 Male   \n",
       "..                                   ...                  ...   \n",
       "183                                adult               Female   \n",
       "184                                adult               Female   \n",
       "185                                adult               Female   \n",
       "186                                adult               Female   \n",
       "187                                adult               Female   \n",
       "\n",
       "    Characteristics[individual]          Characteristics[organism part]  \\\n",
       "0                            D3                    heart left ventricle   \n",
       "1                            D3                    heart left ventricle   \n",
       "2                            D3                    heart left ventricle   \n",
       "3                            D3                    heart left ventricle   \n",
       "4                            D7                    right cardiac atrium   \n",
       "..                          ...                                     ...   \n",
       "183                        AV13  atrioventriculeft cardiac atriumr node   \n",
       "184                        AV13  atrioventriculeft cardiac atriumr node   \n",
       "185                        AV13  atrioventriculeft cardiac atriumr node   \n",
       "186                        AV13  atrioventriculeft cardiac atriumr node   \n",
       "187                        AV13  atrioventriculeft cardiac atriumr node   \n",
       "\n",
       "    Characteristics[disease]  ...          Assay Name   Technology Type  \\\n",
       "0                     normal  ...     HCAHeart9508627  sequencing assay   \n",
       "1                     normal  ...     HCAHeart9508627  sequencing assay   \n",
       "2                     normal  ...     HCAHeart9508627  sequencing assay   \n",
       "3                     normal  ...     HCAHeart9508627  sequencing assay   \n",
       "4                     normal  ...     HCAHeart9508628  sequencing assay   \n",
       "..                       ...  ...                 ...               ...   \n",
       "183                   normal  ...  HCAHeartST13189996  sequencing assay   \n",
       "184                   normal  ...  HCAHeartST13189997  sequencing assay   \n",
       "185                   normal  ...  HCAHeartST13189997  sequencing assay   \n",
       "186                   normal  ...  HCAHeartST13189997  sequencing assay   \n",
       "187                   normal  ...  HCAHeartST13189997  sequencing assay   \n",
       "\n",
       "    Comment[ENA_EXPERIMENT]                                   Scan Name  \\\n",
       "0               ERX10811380     HCAHeart9508627_S1_L001_I1_001.fastq.gz   \n",
       "1               ERX10811380     HCAHeart9508627_S1_L001_I2_001.fastq.gz   \n",
       "2               ERX10811380     HCAHeart9508627_S1_L001_R1_001.fastq.gz   \n",
       "3               ERX10811380     HCAHeart9508627_S1_L001_R2_001.fastq.gz   \n",
       "4               ERX10811381     HCAHeart9508628_S1_L001_I1_001.fastq.gz   \n",
       "..                      ...                                         ...   \n",
       "183             ERX10811425  HCAHeartST13189996_S1_L001_R2_001.fastq.gz   \n",
       "184             ERX10811426  HCAHeartST13189997_S1_L001_I1_001.fastq.gz   \n",
       "185             ERX10811426  HCAHeartST13189997_S1_L001_I2_001.fastq.gz   \n",
       "186             ERX10811426  HCAHeartST13189997_S1_L001_R1_001.fastq.gz   \n",
       "187             ERX10811426  HCAHeartST13189997_S1_L001_R2_001.fastq.gz   \n",
       "\n",
       "                   Comment[SUBMITTED_FILE_NAME] Comment[ENA_RUN]  \\\n",
       "0       HCAHeart9508627_S1_L001_I1_001.fastq.gz      ERR11403589   \n",
       "1       HCAHeart9508627_S1_L001_I2_001.fastq.gz      ERR11403589   \n",
       "2       HCAHeart9508627_S1_L001_R1_001.fastq.gz      ERR11403589   \n",
       "3       HCAHeart9508627_S1_L001_R2_001.fastq.gz      ERR11403589   \n",
       "4       HCAHeart9508628_S1_L001_I1_001.fastq.gz      ERR11403590   \n",
       "..                                          ...              ...   \n",
       "183  HCAHeartST13189996_S1_L001_R2_001.fastq.gz      ERR11403634   \n",
       "184  HCAHeartST13189997_S1_L001_I1_001.fastq.gz      ERR11403635   \n",
       "185  HCAHeartST13189997_S1_L001_I2_001.fastq.gz      ERR11403635   \n",
       "186  HCAHeartST13189997_S1_L001_R1_001.fastq.gz      ERR11403635   \n",
       "187  HCAHeartST13189997_S1_L001_R2_001.fastq.gz      ERR11403635   \n",
       "\n",
       "                                    Comment[FASTQ_URI] Comment[read_index]  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index1   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index2   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read1   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read2   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index1   \n",
       "..                                                 ...                 ...   \n",
       "183  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read2   \n",
       "184  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index1   \n",
       "185  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index2   \n",
       "186  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read1   \n",
       "187  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read2   \n",
       "\n",
       "           Comment[read_type]             Factor Value[organism part]  \n",
       "0              sample_barcode                    heart left ventricle  \n",
       "1              sample_barcode                    heart left ventricle  \n",
       "2    cell_barcode,umi_barcode                    heart left ventricle  \n",
       "3                      single                    heart left ventricle  \n",
       "4              sample_barcode                    right cardiac atrium  \n",
       "..                        ...                                     ...  \n",
       "183                    single  atrioventriculeft cardiac atriumr node  \n",
       "184            sample_barcode  atrioventriculeft cardiac atriumr node  \n",
       "185            sample_barcode  atrioventriculeft cardiac atriumr node  \n",
       "186  cell_barcode,umi_barcode  atrioventriculeft cardiac atriumr node  \n",
       "187                    single  atrioventriculeft cardiac atriumr node  \n",
       "\n",
       "[188 rows x 51 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = pd.read_csv(os.path.join(data_path, meta_data_file), sep='\\t')\n",
    "meta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[FASTQ_URI]</th>\n",
       "      <th>Comment[SUBMITTED_FILE_NAME]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508627_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508627_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508628_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>188 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Source Name                                 Comment[FASTQ_URI]  \\\n",
       "0       HCAHeart9508627  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "1       HCAHeart9508627  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "2       HCAHeart9508627  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "3       HCAHeart9508627  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "4       HCAHeart9508628  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "..                  ...                                                ...   \n",
       "183  HCAHeartST13189996  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "184  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "185  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "186  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "187  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "\n",
       "                   Comment[SUBMITTED_FILE_NAME]  \n",
       "0       HCAHeart9508627_S1_L001_I1_001.fastq.gz  \n",
       "1       HCAHeart9508627_S1_L001_I2_001.fastq.gz  \n",
       "2       HCAHeart9508627_S1_L001_R1_001.fastq.gz  \n",
       "3       HCAHeart9508627_S1_L001_R2_001.fastq.gz  \n",
       "4       HCAHeart9508628_S1_L001_I1_001.fastq.gz  \n",
       "..                                          ...  \n",
       "183  HCAHeartST13189996_S1_L001_R2_001.fastq.gz  \n",
       "184  HCAHeartST13189997_S1_L001_I1_001.fastq.gz  \n",
       "185  HCAHeartST13189997_S1_L001_I2_001.fastq.gz  \n",
       "186  HCAHeartST13189997_S1_L001_R1_001.fastq.gz  \n",
       "187  HCAHeartST13189997_S1_L001_R2_001.fastq.gz  \n",
       "\n",
       "[188 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = meta_data[['Source Name', 'Comment[FASTQ_URI]', 'Comment[SUBMITTED_FILE_NAME]']]\n",
    "meta_data.to_csv('../../raw_data/downloads_table.csv', header=None)\n",
    "meta_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Index file creation\n",
    "\n",
    "**Comprehensive gene annotation ALL:**\n",
    "- It contains the comprehensive gene annotation on the reference chromosomes, scaffolds, assembly patches and alternate loci (haplotypes)\n",
    "\n",
    "**Long non-coding RNA gene annotation CHR:**\n",
    "- It contains the comprehensive gene annotation of lncRNA genes on the reference chromosomes\n",
    "\n",
    "**Genome sequence (GRCh38.p14) ALL:**\n",
    "- Nucleotide sequence of the GRCh38.p14 genome assembly version on all regions, including reference chromosomes, scaffolds, assembly patches and haplotypes\n",
    "- The sequence region names are the same as in the GTF/GFF3 files\n",
    "\n",
    "downloaded from [**here**](https://www.gencodegenes.org/human/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode.v45.chr_patch_hapl_scaff.annotation.gtf.gz\n",
      "gencode.v45.long_noncoding_RNAs.gtf.gz\n",
      "gencode.v45.tRNAs.gtf.gz\n",
      "GRCh38.p14.genome.fa.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "index_input=\"../../reference_data/\"\n",
    "ls $index_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2024-04-11 14:54:36,459]    INFO [ref] Preparing ../../reference_data//GRCh38.p14.genome.fa.gz, ../../reference_data//gencode.v45.chr_patch_hapl_scaff.annotation.gtf.gz\n",
      "[2024-04-11 14:55:24,588]    INFO [ref] Splitting genome ../../reference_data//GRCh38.p14.genome.fa.gz into cDNA at /mnt/LaCIE/ceger/Projects/human_heart_mapping/human_heart_mapping/0-raw_data_processing/tmp/tmpjlu9txpo\n",
      "[2024-04-11 14:56:39,057]    INFO [ref] Concatenating 1 cDNAs to ../../kb_index//cdna.fa\n",
      "[2024-04-11 14:56:39,774]    INFO [ref] Creating transcript-to-gene mapping at ../../kb_index//transcripts_to_genes.txt\n",
      "[2024-04-11 14:56:41,234]    INFO [ref] Indexing ../../kb_index//cdna.fa to ../../kb_index//transcriptome.idx\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "index_input=\"../../reference_data/\"\n",
    "index_output=\"../../kb_index/\"\n",
    "\n",
    "if [ ! -d \"$index_output\" ]; then\n",
    "    mkdir -p \"$index_output\"\n",
    "fi\n",
    "\n",
    "kb ref -i \"$index_output/transcriptome.idx\" \\\n",
    "        -g \"$index_output/transcripts_to_genes.txt\" \\\n",
    "        -f1 \"$index_output/cdna.fa\" \\\n",
    "        \"$index_input/GRCh38.p14.genome.fa.gz\" \\\n",
    "        \"$index_input/gencode.v45.chr_patch_hapl_scaff.annotation.gtf.gz\" \\\n",
    "        \"$index_input/gencode.v45.long_noncoding_RNAs.gtf.gz\" \\\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *optional Preparations before continuing an aborted mapping process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../mapping/HCAHeartST13180619\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "directory=\"../../mapping\"\n",
    "#!/bin/bash\n",
    "\n",
    "\n",
    "# Find folders that don't contain the specific file\n",
    "while IFS= read -r folder; do\n",
    "    if [ ! -f \"$folder/10x_version3_whitelist.txt\" ]; then\n",
    "        echo \"$folder\"\n",
    "        break  # Exit the loop after finding the first folder without the file\n",
    "    fi\n",
    "done < <(find \"$directory\" -mindepth 1 -maxdepth 1 -type d)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make sure that is the correct folder and then delete it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "rm -rf ../../mapping/HCAHeartST13180619"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HCAHeartST13146201',\n",
      "'HCAHeartST10773170',\n",
      "'HCAHeart9845433',\n",
      "'HCAHeartST10773171',\n",
      "'HCAHeartST11064577',\n",
      "'HCAHeartST10773169',\n",
      "'HCAHeartST11350194',\n",
      "'HCAHeartST13180616',\n",
      "'HCAHeart9845436',\n",
      "'HCAHeartST13146206',\n",
      "'HCAHeartST10773166',\n",
      "'HCAHeart9508627',\n",
      "'HCAHeartST11064574',\n",
      "'HCAHeartST11350193',\n",
      "'HCAHeartST13146202',\n",
      "'HCAHeartST11350184',\n",
      "'HCAHeartST13180618',\n",
      "'HCAHeartST11350195',\n",
      "'HCAHeartST11350197',\n",
      "'HCAHeart9845432',\n",
      "'HCAHeartST13146204',\n",
      "'HCAHeartST10773165',\n",
      "'HCAHeartST13146207',\n",
      "'HCAHeartST11064576',\n",
      "'HCAHeart9845434',\n",
      "'HCAHeart9508628',\n",
      "'HCAHeartST13180615',\n",
      "'HCAHeart9845435',\n",
      "'HCAHeartST11350186',\n",
      "'HCAHeart9845431',\n",
      "'HCAHeartST13180617',\n",
      "'HCAHeartST11350198',\n",
      "'HCAHeartST11350192',\n",
      "'HCAHeartST10773168',\n",
      "'HCAHeart9508629',\n",
      "'HCAHeartST13146203',\n",
      "'HCAHeartST13146208',\n",
      "'HCAHeartST13146205',\n",
      "'HCAHeartST11350187',\n",
      "'HCAHeartST10773167',\n",
      "'HCAHeartST11064575',\n",
      "'HCAHeartST13168796']\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Directory to search within\n",
    "directory=\"../../mapping\"\n",
    "\n",
    "# Find folders that contain the specific file\n",
    "# Find folders that contain the specific file\n",
    "folders=$(find \"$directory\" -mindepth 1 -maxdepth 1 -type d -exec test -e \"{}/10x_version3_whitelist.txt\" \\; -printf \"'%f',\\n\")\n",
    "\n",
    "# Remove the trailing comma\n",
    "folders=\"${folders%,}\"\n",
    "\n",
    "# Print the folders in comma-separated format within brackets\n",
    "echo \"[${folders}]\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "copy the output and assign it to a list in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_samples = ['HCAHeartST13146201',\n",
    "'HCAHeartST10773170',\n",
    "'HCAHeart9845433',\n",
    "'HCAHeartST10773171',\n",
    "'HCAHeartST11064577',\n",
    "'HCAHeartST10773169',\n",
    "'HCAHeartST11350194',\n",
    "'HCAHeartST13180616',\n",
    "'HCAHeart9845436',\n",
    "'HCAHeartST13146206',\n",
    "'HCAHeartST10773166',\n",
    "'HCAHeart9508627',\n",
    "'HCAHeartST11064574',\n",
    "'HCAHeartST11350193',\n",
    "'HCAHeartST13146202',\n",
    "'HCAHeartST11350184',\n",
    "'HCAHeartST13180618',\n",
    "'HCAHeartST11350195',\n",
    "'HCAHeartST11350197',\n",
    "'HCAHeart9845432',\n",
    "'HCAHeartST13146204',\n",
    "'HCAHeartST10773165',\n",
    "'HCAHeartST13146207',\n",
    "'HCAHeartST11064576',\n",
    "'HCAHeart9845434',\n",
    "'HCAHeart9508628',\n",
    "'HCAHeartST13180615',\n",
    "'HCAHeart9845435',\n",
    "'HCAHeartST11350186',\n",
    "'HCAHeart9845431',\n",
    "'HCAHeartST13180617',\n",
    "'HCAHeartST11350198',\n",
    "'HCAHeartST11350192',\n",
    "'HCAHeartST10773168',\n",
    "'HCAHeart9508629',\n",
    "'HCAHeartST13146203',\n",
    "'HCAHeartST13146208',\n",
    "'HCAHeartST13146205',\n",
    "'HCAHeartST11350187',\n",
    "'HCAHeartST10773167',\n",
    "'HCAHeartST11064575',\n",
    "'HCAHeartST13168796']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>HCAHeartST11350196</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST11350196_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>HCAHeartST11350196</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST11350196_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>HCAHeartST11350196</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST11350196_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>HCAHeartST11350196</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST11350196_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>HCAHeartST13180619</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13180619_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>HCAHeartST13180619</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13180619_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>HCAHeartST13180619</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13180619_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>HCAHeartST13180619</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13180619_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>HCAHeartST13189995</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189995_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>HCAHeartST13189995</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189995_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>HCAHeartST13189995</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189995_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>HCAHeartST13189995</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189995_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189996_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeartST13189997_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      1                                                  2  \\\n",
       "0                                                                            \n",
       "108  HCAHeartST11350196  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "109  HCAHeartST11350196  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "110  HCAHeartST11350196  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "111  HCAHeartST11350196  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "172  HCAHeartST13180619  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "173  HCAHeartST13180619  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "174  HCAHeartST13180619  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "175  HCAHeartST13180619  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "176  HCAHeartST13189995  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "177  HCAHeartST13189995  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "178  HCAHeartST13189995  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "179  HCAHeartST13189995  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "180  HCAHeartST13189996  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "181  HCAHeartST13189996  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "182  HCAHeartST13189996  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "183  HCAHeartST13189996  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "184  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "185  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "186  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "187  HCAHeartST13189997  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "\n",
       "                                              3  \n",
       "0                                                \n",
       "108  HCAHeartST11350196_S1_L001_I1_001.fastq.gz  \n",
       "109  HCAHeartST11350196_S1_L001_I2_001.fastq.gz  \n",
       "110  HCAHeartST11350196_S1_L001_R1_001.fastq.gz  \n",
       "111  HCAHeartST11350196_S1_L001_R2_001.fastq.gz  \n",
       "172  HCAHeartST13180619_S1_L001_I1_001.fastq.gz  \n",
       "173  HCAHeartST13180619_S1_L001_I2_001.fastq.gz  \n",
       "174  HCAHeartST13180619_S1_L001_R1_001.fastq.gz  \n",
       "175  HCAHeartST13180619_S1_L001_R2_001.fastq.gz  \n",
       "176  HCAHeartST13189995_S1_L001_I1_001.fastq.gz  \n",
       "177  HCAHeartST13189995_S1_L001_I2_001.fastq.gz  \n",
       "178  HCAHeartST13189995_S1_L001_R1_001.fastq.gz  \n",
       "179  HCAHeartST13189995_S1_L001_R2_001.fastq.gz  \n",
       "180  HCAHeartST13189996_S1_L001_I1_001.fastq.gz  \n",
       "181  HCAHeartST13189996_S1_L001_I2_001.fastq.gz  \n",
       "182  HCAHeartST13189996_S1_L001_R1_001.fastq.gz  \n",
       "183  HCAHeartST13189996_S1_L001_R2_001.fastq.gz  \n",
       "184  HCAHeartST13189997_S1_L001_I1_001.fastq.gz  \n",
       "185  HCAHeartST13189997_S1_L001_I2_001.fastq.gz  \n",
       "186  HCAHeartST13189997_S1_L001_R1_001.fastq.gz  \n",
       "187  HCAHeartST13189997_S1_L001_R2_001.fastq.gz  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "download_resume = pd.read_csv('../../raw_data/downloads_table.csv', index_col=0, header=None)\n",
    "download_resume = download_resume[~download_resume[1].isin(mapped_samples)]\n",
    "download_resume"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "download_resume.to_csv('../../raw_data/downloads_table_resume.csv', header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping\n",
    "This script loops over each line from the csv file created above, extracts the sample name, download url and file name from it and creates a folder named after the sample. \\\n",
    "Then it will download the 4 files associated with the sample and once that is done begin processing it with 'kb count'. \\\n",
    "Once that is done the downloaded files are deleted and only the mapping output remains in the sample folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "metadata='../../raw_data/downloads_table_resume.csv'\n",
    "output_path='../../mapping/'\n",
    "index_output=\"../../kb_index/\"\n",
    "\n",
    "# Check if the file exists\n",
    "if [ ! -f \"$metadata\" ]; then\n",
    "    echo \"File not found: $metadata\"\n",
    "    exit 1\n",
    "fi\n",
    "\n",
    "# Read the CSV file and download files for each sample\n",
    "while IFS=, read -r index sample url file_name; do\n",
    "    # Create the directory for the sample\n",
    "    mkdir -p \"$output_path/$sample\"\n",
    "    \n",
    "    # Download the file using axel\n",
    "    axel -n 10 --output=\"$output_path/$sample/$file_name\" \"$url\"\n",
    "    \n",
    "    # Keep track of the downloaded files\n",
    "    downloaded_files+=(\"$output_path/$sample/$file_name\")\n",
    "    \n",
    "    # Check if 4 files have been downloaded for the current sample\n",
    "    if [ ${#downloaded_files[@]} -eq 4 ]; then\n",
    "        echo \"Processing files for sample: $sample\"\n",
    "        \n",
    "        # Run kb count command to process the downloaded files\n",
    "        kb count -i \"$index_output/transcriptome.idx\" -g \"$index_output/transcripts_to_genes.txt\" -x 10XV3 \\\n",
    "            --h5ad \\\n",
    "            \"$output_path/$sample\"/*.fastq.gz \\\n",
    "            -o \"$output_path/$sample/\"\n",
    "        \n",
    "        echo \"Deleting downloaded files for sample: $sample\"\n",
    "        # Delete only the downloaded files\n",
    "        rm \"${downloaded_files[@]}\"\n",
    "        # Clear the downloaded files array for the next sample\n",
    "        downloaded_files=()\n",
    "    fi\n",
    "done < \"$metadata\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "heart_mapping",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
