{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for mapping ATACseq Data\n",
    "\n",
    "Raw data downloaded from [**here**](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-12919/sdrf)\n",
    "\n",
    "    Developed by: Christian Eger\n",
    "    Würzburg Institute for Systems Immunology - Faculty of Medicine - Julius Maximilian Universität Würzburg\n",
    "    Created on: 240415\n",
    "    Last modified: 240416"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating input file for download and processing script\n",
    "\n",
    "Downloaded from [**here**](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-12919)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../.data/meta_data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[age]</th>\n",
       "      <th>Characteristics[developmental stage]</th>\n",
       "      <th>Characteristics[sex]</th>\n",
       "      <th>Characteristics[individual]</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Characteristics[disease]</th>\n",
       "      <th>...</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Technology Type</th>\n",
       "      <th>Comment[ENA_EXPERIMENT]</th>\n",
       "      <th>Scan Name</th>\n",
       "      <th>Comment[SUBMITTED_FILE_NAME]</th>\n",
       "      <th>Comment[ENA_RUN]</th>\n",
       "      <th>Comment[FASTQ_URI]</th>\n",
       "      <th>Comment[read_index]</th>\n",
       "      <th>Comment[read_type]</th>\n",
       "      <th>Factor Value[organism part]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ERS15408182</td>\n",
       "      <td>SAMEA113413051</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811516</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>ERR11403725</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index1</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ERS15408182</td>\n",
       "      <td>SAMEA113413051</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811516</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>ERR11403725</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index2</td>\n",
       "      <td>cell_barcode</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ERS15408182</td>\n",
       "      <td>SAMEA113413051</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811516</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>ERR11403725</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read1</td>\n",
       "      <td>paired</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ERS15408182</td>\n",
       "      <td>SAMEA113413051</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>55 to 60</td>\n",
       "      <td>adult</td>\n",
       "      <td>male</td>\n",
       "      <td>D3</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811516</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>ERR11403725</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>read2</td>\n",
       "      <td>paired</td>\n",
       "      <td>heart left ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>ERS15408183</td>\n",
       "      <td>SAMEA113413052</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>60 to 65</td>\n",
       "      <td>adult</td>\n",
       "      <td>male</td>\n",
       "      <td>D7</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>normal</td>\n",
       "      <td>...</td>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>sequencing assay</td>\n",
       "      <td>ERX10811517</td>\n",
       "      <td>HCAHeart9508820_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>HCAHeart9508820_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>ERR11403726</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>index1</td>\n",
       "      <td>sample_barcode</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0  HCAHeart9508819         ERS15408182        SAMEA113413051   \n",
       "1  HCAHeart9508819         ERS15408182        SAMEA113413051   \n",
       "2  HCAHeart9508819         ERS15408182        SAMEA113413051   \n",
       "3  HCAHeart9508819         ERS15408182        SAMEA113413051   \n",
       "4  HCAHeart9508820         ERS15408183        SAMEA113413052   \n",
       "\n",
       "  Characteristics[organism] Characteristics[age]  \\\n",
       "0              Homo sapiens             55 to 60   \n",
       "1              Homo sapiens             55 to 60   \n",
       "2              Homo sapiens             55 to 60   \n",
       "3              Homo sapiens             55 to 60   \n",
       "4              Homo sapiens             60 to 65   \n",
       "\n",
       "  Characteristics[developmental stage] Characteristics[sex]  \\\n",
       "0                                adult                 male   \n",
       "1                                adult                 male   \n",
       "2                                adult                 male   \n",
       "3                                adult                 male   \n",
       "4                                adult                 male   \n",
       "\n",
       "  Characteristics[individual] Characteristics[organism part]  \\\n",
       "0                          D3           heart left ventricle   \n",
       "1                          D3           heart left ventricle   \n",
       "2                          D3           heart left ventricle   \n",
       "3                          D3           heart left ventricle   \n",
       "4                          D7           right cardiac atrium   \n",
       "\n",
       "  Characteristics[disease]  ...       Assay Name   Technology Type  \\\n",
       "0                   normal  ...  HCAHeart9508819  sequencing assay   \n",
       "1                   normal  ...  HCAHeart9508819  sequencing assay   \n",
       "2                   normal  ...  HCAHeart9508819  sequencing assay   \n",
       "3                   normal  ...  HCAHeart9508819  sequencing assay   \n",
       "4                   normal  ...  HCAHeart9508820  sequencing assay   \n",
       "\n",
       "  Comment[ENA_EXPERIMENT]                                Scan Name  \\\n",
       "0             ERX10811516  HCAHeart9508819_S1_L001_I1_001.fastq.gz   \n",
       "1             ERX10811516  HCAHeart9508819_S1_L001_I2_001.fastq.gz   \n",
       "2             ERX10811516  HCAHeart9508819_S1_L001_R1_001.fastq.gz   \n",
       "3             ERX10811516  HCAHeart9508819_S1_L001_R2_001.fastq.gz   \n",
       "4             ERX10811517  HCAHeart9508820_S1_L001_I1_001.fastq.gz   \n",
       "\n",
       "              Comment[SUBMITTED_FILE_NAME] Comment[ENA_RUN]  \\\n",
       "0  HCAHeart9508819_S1_L001_I1_001.fastq.gz      ERR11403725   \n",
       "1  HCAHeart9508819_S1_L001_I2_001.fastq.gz      ERR11403725   \n",
       "2  HCAHeart9508819_S1_L001_R1_001.fastq.gz      ERR11403725   \n",
       "3  HCAHeart9508819_S1_L001_R2_001.fastq.gz      ERR11403725   \n",
       "4  HCAHeart9508820_S1_L001_I1_001.fastq.gz      ERR11403726   \n",
       "\n",
       "                                  Comment[FASTQ_URI] Comment[read_index]  \\\n",
       "0  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index1   \n",
       "1  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index2   \n",
       "2  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read1   \n",
       "3  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...               read2   \n",
       "4  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...              index1   \n",
       "\n",
       "  Comment[read_type] Factor Value[organism part]  \n",
       "0     sample_barcode        heart left ventricle  \n",
       "1       cell_barcode        heart left ventricle  \n",
       "2             paired        heart left ventricle  \n",
       "3             paired        heart left ventricle  \n",
       "4     sample_barcode        right cardiac atrium  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = pd.read_csv(\n",
    "    filepath_or_buffer=os.path.join(data_path, 'E-MTAB-12919.sdrf.txt'),\n",
    "    sep='\\t',\n",
    ")\n",
    "meta_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[FASTQ_URI]</th>\n",
       "      <th>Comment[SUBMITTED_FILE_NAME]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508819_S1_L001_I2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508819_S1_L001_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...</td>\n",
       "      <td>HCAHeart9508820_S1_L001_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Source Name                                 Comment[FASTQ_URI]  \\\n",
       "0  HCAHeart9508819  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "1  HCAHeart9508819  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "2  HCAHeart9508819  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "3  HCAHeart9508819  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "4  HCAHeart9508820  ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR114/ERR114...   \n",
       "\n",
       "              Comment[SUBMITTED_FILE_NAME]  \n",
       "0  HCAHeart9508819_S1_L001_I1_001.fastq.gz  \n",
       "1  HCAHeart9508819_S1_L001_I2_001.fastq.gz  \n",
       "2  HCAHeart9508819_S1_L001_R1_001.fastq.gz  \n",
       "3  HCAHeart9508819_S1_L001_R2_001.fastq.gz  \n",
       "4  HCAHeart9508820_S1_L001_I1_001.fastq.gz  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = meta_data[['Source Name', 'Comment[FASTQ_URI]', 'Comment[SUBMITTED_FILE_NAME]']]\n",
    "meta_data.to_csv('../.data/meta_data/downloads_table.csv', header=None)\n",
    "meta_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[**Documentation**](https://github.com/pachterlab/scATAK/tree/main)\n",
    "\n",
    "### Usage\n",
    "```\n",
    "$SCATAK_HOME/scATAK -h\n",
    "\n",
    "######################################################################################\n",
    "\n",
    "scATAK [options]\n",
    "-module=quant [please choose 'quant' for single-cell quantification, 'track' for group bigwig track generation, 'hic' for HiC related analysis]\n",
    "Please specify the following options:\n",
    "-id --sample_id=sample_sheet.csv [a sample information sheet for fastq files, must be csv format]\n",
    "-genome --genome_fasta=Mus_musculus.GRCm38.dna_rm.primary_assembly.fa [ENSEMBL genome fasta file for the organism of your interest]\n",
    "-gene --gene_gtf=Mus_musculus.GRCm38.101.chr.gtf [ENSEMBL gene gtf file for the organism of your interest]\n",
    "-bc --blen=16 [length of cell barcode, default:16]\n",
    "-bf --flen=40 [length of biological feature, should not be longer than R2 read length, default:40]\n",
    "-bg --bc_group=bc_group.txt [a two-column text file with Barcode and Group. First line should be 'Barcode' and 'Group'] (for -module=track or -module=hic\n",
    "-bam --bam_file=peak_calling/sampleX.bam [scATAK quant mapped bam file for sampleX] (for -module=track only)\n",
    "-hic --hic_bedpe=sample_significant.bedpe [hic interaction bedpe file] (for -module=hic only)\n",
    "-bin --hic_binsize=10000 [hic interaction bin size, default 10kb] (for -module=hic only)\n",
    "-mtxdir --region_mtxdir=atac_regions/atac_sampleX [atac region matrix directory for sampleX] (for -module=hic only)\n",
    "-t --thread=8 [threads to use, default:8]\n",
    "-h --help [Help information]\n",
    "```\n",
    "### Input File Format\n",
    "#### sample_sheet.csv\n",
    "sample ID \tSample Group \tR1 Biological Read FASTQ \tR2 Cell Barcode Read FASTQ \tR3 Biological Read FASTQ \tBarcode Whitelist\n",
    "```\n",
    "male1,male,B8_S1_L001_R1_001.fastq.gz,B8_S1_L001_R2_001.fastq.gz,B8_S1_L001_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "male2,male,B9_S1_L003_R1_001.fastq.gz,B9_S1_L003_R2_001.fastq.gz,B9_S1_L003_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "male3,male,B10_S1_L004_R1_001.fastq.gz,B10_S1_L004_R2_001.fastq.gz,B10_S1_L004_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "female1,female,4_F_S1_L004_R1_001.fastq.gz,4_F_S1_L004_R2_001.fastq.gz,4_F_S1_L004_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "female2,female,5_F_S1_L001_R1_001.fastq.gz,5_F_S1_L001_R2_001.fastq.gz,5_F_S1_L001_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "female3,female,6_F_S1_L002_R1_001.fastq.gz,6_F_S1_L002_R2_001.fastq.gz,6_F_S1_L002_R3_001.fastq.gz,/home/fgao/scATAK/lib/737K-cratac-v1.txt\n",
    "```\n",
    "#### bc_group.txt\n",
    "Cell Barcodes \tCell Type Group\n",
    "```\n",
    "Barcodes Groups\n",
    "AAACGAAAGCCTCGCA Oligodendrocytes\n",
    "AAACGAAAGGAAGAAC Oligodendrocytes\n",
    "AAACGAACAGCAACGA OPC\n",
    "AAACGAACATTACTCT Microglia\n",
    "AAACGAATCACTCGGG Oligodendrocytes\n",
    "AAACGAATCCTTACGC Oligodendrocytes\n",
    "AAACGAATCGATCTTT Astrocytes\n",
    "```\n",
    "### Quick Start\n",
    "```\n",
    "#Quant module\n",
    "$SCATAK_HOME/scATAK -module=quant -id=sample_sheet.csv -genome=Homo_sapiens.GRCh38.dna_rm.primary_assembly.fa -gene=Homo_sapiens.GRCh38.101.chr.gtf -bc=16 -bf=40 -t=4\n",
    "\n",
    "#Track module\n",
    "$SCATAK_HOME/scATAK -module=track -bg=bc_group.txt -bam=peak_calling/sampleX.bam -genome=Homo_sapiens.GRCh38.dna_rm.primary_assembly.fa\n",
    "\n",
    "#Hic module\n",
    "$SCATAK_HOME/scATAK -module=hic -bg=bc_group.txt -hic=sample_significant.bedpe -bin=10000 -mtxdir=atac_regions/atac_sampleX -t=16\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "metadata='../.data/meta_data/downloads_table.csv'\n",
    "output_path='../.data/mapping/'\n",
    "index_output=\"../.data/kb_index/\"\n",
    "\n",
    "if [ ! -f \"$metadata\" ]; then\n",
    "    echo \"File not found: $metadata\"\n",
    "    exit 1\n",
    "fi\n",
    "\n",
    "while IFS=, read -r index sample url file_name; do\n",
    "    mkdir -p \"$output_path/$sample\"\n",
    "    \n",
    "    axel -n 10 --output=\"$output_path/$sample/$file_name\" \"$url\"\n",
    "    \n",
    "    downloaded_files+=(\"$output_path/$sample/$file_name\")\n",
    "    \n",
    "    if [ ${#downloaded_files[@]} -eq 4 ]; then\n",
    "        echo \"Processing files for sample: $sample\"\n",
    "        \n",
    "        echo \"Deleting downloaded files for sample: $sample\"\n",
    "        #rm \"${downloaded_files[@]}\"\n",
    "        downloaded_files=()\n",
    "    fi\n",
    "done < \"$metadata\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "heart_mapping",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
